certbot_route53/authenticator.py

```diff
@@ -14,112 +16,128 @@
-            ),
-            key=lambda zone: len(zone["Name"]),
-        )
-
-
-    def _perform_single(self, achall):
-        # provision the TXT record, using the domain name given. Assumes the hosted zone exits, else fails the challenge
-        r53 = boto3.client('route53')
-        logger.info("Doing validation for " + achall.domain)
-
         try:
-            zone = self._find_zone(r53, achall.domain)
-        except ValueError as e:
-            logger.error("Unable to find matching Route53 zone for domain " + achall.domain)
-            return None
-
-        response, validation = achall.response_and_validation()
-        self._excute_r53_action(r53, achall, zone, validation, 'UPSERT', wait_for_change=True)
-
-        for _ in range(TTL*2):
-            if response.simple_verify(
-                achall.chall,
-                achall.domain,
-                achall.account_key.public_key(),
-            ):
-                break
-            logger.info("Waiting for DNS propagation...")
-            time.sleep(1)
-        else:
-            logger.error("Unable to verify domain " + achall.domain)
-            return None
-
-        return response
-
-    def cleanup(self, achalls):
-        # pylint: disable=missing-docstring
-        r53 = boto3.client('route53')
-        for achall in achalls:
-            try:
-                zone = self._find_zone(r53, achall.domain)
-            except ValueError:
-                logger.warn("Unable to find zone for " + achall.domain + ". Skipping cleanup.")
-                continue
-
-            _, validation = achall.response_and_validation()
-            self._excute_r53_action(r53, achall, zone, validation, 'DELETE')
-        return None
-
-
-    def _excute_r53_action(self, r53, achall, zone, validation, action, wait_for_change=False):
-            response = r53.change_resource_record_sets(
-                HostedZoneId=zone["Id"],
-                ChangeBatch={
-                    'Comment': 'Let\'s Encrypt ' + action,
-                    'Changes': [
-                        {
-                            'Action': action,
-                            'ResourceRecordSet': {
-                                'Name': achall.validation_domain_name(achall.domain),
-                                'Type': 'TXT',
-                                'TTL': TTL,
-                                'ResourceRecords': [
-                                    {
-                                        'Value': '"' + validation + '"',
-                                    },
-                                ],
-                            },
-                        },
-                    ],
-                },
+            change_ids = [self._create_single(achall) for achall in achalls]
+            for change_id in change_ids:
+                self._wait_for_change(change_id)
+            return [achall.response(achall.account_key) for achall in achalls]
```

Comment:
> The external verification is requested by Certbot once the perform call returns. I believe internal verification isn't necessary because Route53 provide the handy INSYNC API response.

I'm not sure how these two are related. But, then again, I'm not really sure what's happening with the whole `achalls` business. From your answer, I'm guessing that things Just Happen™ after you return and if you've put the right values in the right TXT records.

I'll give it a spin and complain if things don't work.